subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -529,242 +599,209 @@ func (h *HivedAlgorithm) findVirtualCell(
-		}
+// generatePodScheduleResult writes the scheduling result into a PodScheduleResult.
+func generatePodScheduleResult(
+	groupPhysicalPlacement map[int32][]CellList,
+	groupVcPlacement map[int32][]CellList,
+	podIndex int32,
+	suggestedNodes []string,
+	pod *core.Pod,
+	newGroup bool) internal.PodScheduleResult {
 
-		// set priority
-		setPriority(virtualCell, priority, unlimitedLevel)
-		// first hold the physical cell of the preassigned (set its priority to regularPriority)
-		if !preassignedBound {
-			setPriority(pac.GetPhysicalCell(), regularPriority, unlimitedLevel)
-		}
-		// set the priority of the physical cell to regularNonPreassignedPriority
-		// so that other cells inside the preassigned one can still be used by opportunistic pods
-		if virtualCell != pac {
-			setSelfAndParentPriority(physicalCell, regularNonPreassignedPriority, unlimitedLevel)
+	if groupPhysicalPlacement == nil {
+		return internal.PodScheduleResult{
+			PodWaitInfo: &internal.PodWaitInfo{
+				FailedNodeReasons: map[string]string{},
+			},
 		}
 	} else {
-		setPriority(physicalCell, opportunisticPriority, unlimitedLevel)
-	}
-}
-
-// confirmReleasedCell releases an allocated cell by destroying cell bindings and reset priorities.
-func confirmReleasedCell(c *PhysicalCell) {
-	var preassignedPhysical *PhysicalCell
-	if vc := c.GetVirtualCell(); vc != nil {
-		// destroy bindings
-		boundVirtual := vc
-		for !boundVirtual.GetPhysicalCell().IsReserved() {
-			boundPhysical := boundVirtual.GetPhysicalCell()
-			klog.Infof("Cells unbound: %v and %v", boundVirtual.GetName(), boundPhysical.GetName())
-			boundPhysical.SetVirtualCell(nil)
-			boundVirtual.SetPhysicalCell(nil)
-			if boundVirtual.GetParent() == nil {
-				break
-			} else {
-				unbindParent := true
-				for _, cc := range boundVirtual.GetParent().GetChildren() {
-					if child := cc.(*VirtualCell); child.GetPhysicalCell() != nil {
-						unbindParent = false
-						break
+		var (
+			selectedNode       string
+			selectedGpuIndices []int32
+			chain              string
+		)
+		preemptionVictims := common.NewSet()
+		gsi := make([]api.GroupMemberScheduleInfo, len(groupPhysicalPlacement))
+		groupMemberIndex := 0
+		for podGpuNum, podPhysicalPlacements := range groupPhysicalPlacement {
+			gms := api.GroupMemberScheduleInfo{
+				GpuNumber:     podGpuNum,
+				PodPlacements: make([]api.PodPlacementInfo, len(podPhysicalPlacements)),
+			}
+			for i := int32(0); i < int32(len(podPhysicalPlacements)); i++ {
+				gms.PodPlacements[i].PhysicalGpuIndices = make(
+					[]int32, len(podPhysicalPlacements[i]))
+				gms.PodPlacements[i].VirtualGpuCellIndices = make(
+					[]int32, len(podPhysicalPlacements[i]))
+				for j := 0; j < len(podPhysicalPlacements[i]); j++ {
+					pGpu := podPhysicalPlacements[i][j].(*PhysicalCell)
+					nodes, gpuIndices := pGpu.GetPhysicalPlacement()
+					gms.PodPlacements[i].PhysicalNode = nodes[0]
+					gms.PodPlacements[i].PhysicalGpuIndices[j] = gpuIndices[0]
```

Comment:
>gpuIndices[0] [](start = 50, length = 13)

Why only choose the first gpu?

Reply:
Or they must be one gpu cells? So do we have to reuse the cell structure? It may introduce confuse, why not use gpu directly.

---
In reply to: [332835528](https://github.com/microsoft/pai/pull/3684#discussion_r332835528) [](ancestors = 332835528)